dags/top-public-synapse-projects-30-days-from-snowflake.py

```diff
@@ -0,0 +1,170 @@
+from airflow.models.param import Param
+from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+from orca.services.synapse import SynapseHook
+
+dag_params = {
+    "snowflake_conn_id": Param("SNOWFLAKE_SYSADMIN_PORTAL_RAW_CONN", type="string"),
+    "synapse_conn_id": Param("SYNAPSE_ORCA_SERVICE_ACCOUNT_CONN", type="string"),
+}
+
+dag_config = {
+    "schedule_interval": "0 0 * * *",
+    "start_date": datetime(2024, 4, 1),
+    "catchup": False,
+    "default_args": {
+        "retries": 1,
+    },
+    "tags": ["snowflake"],
+    "params": dag_params,
+}
+
+SYNAPSE_RESULTS_TABLE = "syn55382267"
+
+
+@dataclass
+class DownloadMetric:
+    """Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        name: The name of the project
+        project: The ID of the project
+        downloads_per_project: The number of downloads per project
+        number_of_unique_users_downloaded: The number of unique users who downloaded
+        data_download_size: The size of the data downloaded
+
+    """
+
+    name: str
+    project: str
+    downloads_per_project: int
+    number_of_unique_users_downloaded: int
+    data_download_size: float
+
+
+@dag(**dag_config)
+def top_public_synapse_projects_30_days_from_snowflake() -> None:
+    """Execute a query on Snowflake and report the results to a slack channel and
+    synapse table."""
+
+    @task
+    def get_all_time_downloads_from_snowflake(**context) -> List[DownloadMetric]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = """
+            WITH PUBLIC_PROJECTS AS (
+                SELECT
+                    node_latest.project_id,
+                    node_latest.name
+                FROM
+                    synapse_data_warehouse.synapse.node_latest
+                WHERE
+                    node_latest.is_public AND
+                    node_latest.node_type = 'project'
+            ),
+            DEDUP_FILEHANDLE AS (
+                SELECT DISTINCT
+                    PUBLIC_PROJECTS.name,
+                    filedownload.user_id,
+                    filedownload.file_handle_id AS FD_FILE_HANDLE_ID,
+                    filedownload.record_date,
+                    filedownload.project_id,
+                    file_latest.content_size
+                FROM
+                    synapse_data_warehouse.synapse.filedownload
+                INNER JOIN
+                    PUBLIC_PROJECTS
+                ON
+                    filedownload.project_id = PUBLIC_PROJECTS.project_id
+                INNER JOIN
+                    synapse_data_warehouse.synapse.file_latest
+                ON
+                    filedownload.file_handle_id = file_latest.id
+                WHERE
+                    filedownload.record_date > DATEADD(DAY, -30, CURRENT_DATE)
```

Comment:
It might be worthwhile to allow "current_date" to be passed into this DAG. If this misses a day and we want to back-fill the data for the day that was missed we currently do not have a way of doing this.

If we passed the configuration into the DAG we could execute it for any arbitrary 30 day period. It would need to be a change here, and the today var you defined in `push_results_to_synapse_table`

Reply:
I suppose the query would also need to be updated to only select those records `<= current date` and `> -30 days`

Reply:
I updated the DAG to have a parameterized `CURRENT_DATE`. 

> I suppose the query would also need to be updated to only select those records `<= current date` and `> -30 days`

Isn't that what I am already [doing](https://github.com/Sage-Bionetworks-Workflows/orca-recipes/blob/0f4324c58afd27c19bfdb3a275b8272c19b06d17/dags/top-public-synapse-projects-30-days-from-snowflake.py#L95-L98)?

Reply:
> I updated the DAG to have a parameterized `CURRENT_DATE`.
> 
> > I suppose the query would also need to be updated to only select those records `<= current date` and `> -30 days`
> 
> Isn't that what I am already [doing](https://github.com/Sage-Bionetworks-Workflows/orca-recipes/blob/0f4324c58afd27c19bfdb3a275b8272c19b06d17/dags/top-public-synapse-projects-30-days-from-snowflake.py#L95-L98)?

Not quite - In your query you are looking at all records that are: `filedownload.record_date > DATEADD(DAY, -30, '{context["params"]["current_date"]}')`

Suppose you run this query for last month (03/09/2024). Your query would answer the question:
"Get me all filedownloads that have a record date greater than 02/09/2024"

The data you are retrieving is going to include 2 months worth of data since it would include everything also collected between 03/09/2024 -> 04/09/2024.

By including a `<= current date` we are setting a lower (-30 days) and an upper (today) bound on the data being selected.

Reply:
Ignore my comment - I looked back at the code and saw you are already doing this. I missed the `AND`:
```
                WHERE
                    filedownload.record_date > DATEADD(DAY, -30, '{context["params"]["current_date"]}')
                AND 
                    filedownload.record_date <= '{context["params"]["current_date"]}'
```